simpeg/potential_fields/magnetics/simulation.py

```diff
@@ -1730,497 +1800,387 @@ def getB0(self):
-        dCdm_RHS1 = Div * (sdiag(self.MfMu0 * B0) * dMfMuI)
-        # temp1 = (Dface * (self._Pout.T * self.Bbc_const * self.Bbc))
-        # dCdm_RHS2v = (sdiag(vol) * temp1) * \
-        #    np.inner(vol, dchidmu * dmu_dm * v)
-
-        # dCdm_RHSv =  dCdm_RHS1*(dmu_dm*v) +  dCdm_RHS2v
-        dCdm_RHSv = dCdm_RHS1 * (dmu_dm * v)
-        dCdm_v = dCdm_A * v - dCdm_RHSv
-
-        Ainv = self.solver(dCdu, **self.solver_opts)
-        sol = Ainv * dCdm_v
-
-        dudm = -sol
-        dBdmv = (
-            sdiag(self.MfMu0 * B0) * (dMfMuI * (dmu_dm * v))
-            - sdiag(Div.T * u) * (dMfMuI * (dmu_dm * v))
-            - self.MfMuI * (Div.T * (dudm))
-        )
+        if f is None:
+            f = self.fields(m)
 
-        Ainv.clean()
+        if self.storeJ:
+            J = self.getJ(m, f=f)
+            return J.dot(v)
 
-        return mkvc(P * dBdmv)
+        return self._Jvec(m, v, f)
 
-    @utils.timeIt
-    def Jtvec(self, m, v, u=None):
-        r"""
-        Computing Jacobian^T multiplied by vector.
+    def Jtvec(self, m, v, f=None):
+        self.model = m
 
-        .. math ::
+        if f is None:
+            f = self.fields(m)
 
-            (\frac{\delta \mathbf{P}\mathbf{B}} {\delta \mathbf{m}})^{T} =
-                \left[
-                    \mathbf{P}_{deriv}\frac{\partial \mathbf{\mu} } {\partial \mathbf{m} }
-                    \left[
-                        \diag(\M^f_{\mu_{0}^{-1} } \mathbf{B}_0) \dMfMuI
-                         - \diag (\Div^T\mathbf{u})\dMfMuI
-                    \right ]
-                \right]^{T}
-                 -
-                 \left[
-                     \mathbf{P}_{deriv}(\MfMui)^{-1} \Div^T
-                     \frac{\delta\mathbf{u}}{\delta \mathbf{m}}
-                 \right]^{T}
+        if self.storeJ:
+            J = self.getJ(m, f=f)
+            return np.asarray(J.T.dot(v))
 
-        where
+        return self._Jtvec(m, v, f)
 
-        .. math ::
+    def getJ(self, m, f=None):
+        self.model = m
+        if self._Jmatrix is None:
+            if f is None:
+                f = self.fields(m)
+            if m.size < self.survey.nD:
+                J = self._Jvec(m, v=None, f=f)
+            else:
+                J = self._Jtvec(m, v=None, f=f).T
 
-            \mathbf{P}_{derv} = \frac{\partial \mathbf{P}}{\partial\mathbf{B}}
+        else:
+            J = self._Jmatrix
 
-        .. note ::
+        if self.storeJ is True:
+            self._Jmatrix = J
 
-            Here we only want to compute
+        return J
 
-            .. math ::
+    def _Jtvec(self, m, v, f):
+        B, u = f["B"], f["u"]
 
-                \mathbf{J}^{T}\mathbf{v} =
-                (\frac{\delta \mathbf{P}\mathbf{B}} {\delta \mathbf{m}})^{T} \mathbf{v}
+        Q = self._projectFieldsDeriv(B)
 
-        """
-        if u is None:
-            u = self.fields(m)
+        if v is None:
+            v = np.eye(Q.shape[0])
+            DivTatsol_p_QT = (
```

Comment:
```suggestion
            div_solve_q = (
```
Just a suggestion. Those names are getting a bit nuts IMO

Reply:
Tried to simplify the names throughout a bit